### ISSUE TITLE
Lifecycle hooks (methods and events) plus BREAKING CHANGES

### ISSUE THREAD

**Issue by martypdx** [ractivejs/ractive#1256]
PR #1236 added new lifecycle hooks to ractive instances, plus caused a few BREAKING CHANGES (see below).

The hooks are available as methods or events. Methods are listed below, event equivalent would be same without the "on", for example `ractive.oninit` is the method, `ractive.on('init', function(){...})` the event.
##### `onconstruct`

Same as current `beforeInit`, has the constructor `options` as argument. Called before any processing is done. 
##### `onconfig`

Called after all options have been process for the instance based on inheritance and any function options, but viewmodel or virtualDOM has not been created yet. You can access configured options on the ractive instance via `this`.
##### `oninit`

The viewmodel and the virtualDOM (root fragment) have been created, but no rendering has occurred. Only fires once per instance.
##### `onrender`

The view has been rendered. Can be called more than once if instance is unrendered and then re-rendered
##### `onchange`

Existing `change` event
##### `oncomplete`

Existing `complete` event.
##### `onupdate`

Existing `update` event.
##### `onunrender`

New event, companion to `onrender`. Can be called more than once.
##### `onteardown`

Like existing `teardown`, except slight BREAKING CHANGE in that it has been moved to after unrender when teardown causes unrender (which is most of the time).
##### `ondetach`

Fires when `ractive.detach()` is called.
##### `oninsert`

Fires when `ractive.insert()` is called.
#### notes

`onconstruct`, `onconfig`, `oninit` and `onrender` are top-down events so in component hierarchies parents will fire before children. The rest are bottom-up and will fire in each component as the event happens.

Old methods (`beforeInit`, `init`, and `complete`) are deprecated but supported for the near future. Note there is a potentially BREAKING CHANGE in that the `init` method has been mapped to the `onrender` method. `onrender` is the most equivalent method to the current `init` implementation, except that it will fire more than once and the `options` are no longer provided as an argument.

There are now more reserved event words which may be a BREAKING CHANGE if you use construct, config, render, unrender, detach, or insert.

Another potential BREAKING CHANGE depending on your code is that ractive instances passed options now inherit all of the options and are first-class citizens for hooks and custom methods (`this._super()` fully supported):

``` js
var ractive = new Ractive({
    oninit: function(){
        // hooks without extend!
        this.observe('some.data', function(n,o,k){
            this.mymethod(n)
        })
    },
    mymethod: function(n){
      //methods without extend!
   }
})
```

Hook events bubble, so you can watch components:

``` js
var ractive = new Ractive({
    oninit: function(){
        this.on('widget.render', function(){...})
    },
    components: {
        widget: Ractive.extends()
    }
})
```
